File: 剑指offer/面试题56 - I. 数组中数字出现的次数.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumbers = function(nums) {
  let sum = 0
  //   sum为两个不同值的异或值
  for (let item of nums) {
    sum ^= item
  }
  //   得到二进制1的最低位
  let flag = -sum & sum
  let result = [0, 0]
  //  用flag两组去进行^
  for (let item of nums) {
    if ((flag & item) === 0) {
      result[0] = result[0] ^ item
    } else {
      result[1] = result[1] ^ item
    }
  }
  return result
}

File: Daily/面试题56 - I. 数组中数字出现的次数.py
class Solution:
    def singleNumbers(self, nums: List[int]) -> List[int]:
      ret = 0
      a = 0
      b = 0
      for n in nums:
        ret ^= n
      h = 1
      while(ret & h == 0): 
        h <<= 1
      for n in nums: 
        if (h & n == 0):
          a ^= n 
        else:
          b ^= n 
      return [a, b]